File: services/api/src/api/routes/annotate.py
from fastapi import APIRouter

from api.schemas.annotate import (
    AnnotateRequest,
    AnnotateResponse,
    AnnotateDefinitionData,
    AnnotateUsageData,
)
from babeling_nlp.define import get_definition_candidates
from babeling_nlp.utils import mark_words, extract_text_group
from api.services.gemini import get_gemini


router = APIRouter(prefix="/annotate", tags=["annotate"])


@router.post("")
def annotate(req: AnnotateRequest):
    # -------------------------
    # Extract source / target sentence / paragraph
    # -------------------------
    sent_id = req.tgt.sent_ids[req.tgt_idx]
    par_id = req.tgt.par_ids[req.tgt_idx]

    src_sentence = extract_text_group(
        req.src.words, req.src.spaces, req.src.sent_ids, sent_id
    )
    src_paragraph = extract_text_group(
        req.src.words, req.src.spaces, req.src.par_ids, par_id
    )
    tgt_sentence = extract_text_group(
        req.tgt.words, req.tgt.spaces, req.tgt.sent_ids, sent_id
    )
    tgt_paragraph = extract_text_group(
        req.tgt.words, req.tgt.spaces, req.tgt.par_ids, par_id
    )

    # -------------------------
    # Mark source / target words in sentence
    # -------------------------
    src_idxs = set(req.tgt_to_src.get(req.tgt_idx, []))
    tgt_idxs = {req.tgt_idx}

    src_marked = mark_words(req.src.words, src_idxs, "SOURCE")
    tgt_marked = mark_words(req.tgt.words, tgt_idxs, "TARGET")

    src_marked_sent = extract_text_group(
        src_marked, req.src.spaces, req.src.sent_ids, sent_id
    )
    tgt_marked_sent = extract_text_group(
        tgt_marked, req.tgt.spaces, req.tgt.sent_ids, sent_id
    )

    # -------------------------
    # Get definition candidate info
    # -------------------------
    tgt_word = req.tgt.words[req.tgt_idx]
    candidates = get_definition_candidates(tgt_word, req.tgt_lang)

    gemini_api = get_gemini(req.src_lang, req.tgt_lang)

    res = gemini_api.annotate(
        candidates=candidates,
        tgt_word=tgt_word,
        tgt_lang=req.tgt_lang,
        src_marked_sent=src_marked_sent,
        tgt_marked_sent=tgt_marked_sent,
    )

    definition_data = res["definition"]
    usage_data = res["usage"]

    return AnnotateResponse(
        definition=AnnotateDefinitionData(
            form=definition_data["form"],
            pos_form=definition_data["pos_form"],
            ipa_form=definition_data["ipa_form"],
            lemma=definition_data["lemma"],
            pos_lemma=definition_data["pos_lemma"],
            ipa_lemma=definition_data["ipa_lemma"],
            gloss=definition_data["gloss"],
            src_sentence=src_sentence,
            src_paragraph=src_paragraph,
            tgt_sentence=tgt_sentence,
            tgt_paragraph=tgt_paragraph,
        ),
        usage=AnnotateUsageData(
            explanation=usage_data["explanation"],
            examples=usage_data["examples"],
        ),
    )

File: apps/web/app/(protected)/documents/feature/types/annotate.ts
import { ReaderSession } from "./readerSession";
import { DefineEntry, ExplainEntry } from "../components/Annotate/AnnotateCard";

// -------------------------
// Frontend return types
// -------------------------
export type AnnotateDomainData = {
  words: string[];
  spaces: string[];
  sentIds: number[];
  parIds: number[];
};

export type AnnotateDefinitionData = {
  form: string;
  posForm: string;
  ipaForm: string;

  lemma: string;
  posLemma: string;
  ipaLemma: string;

  gloss: string;

  srcSentence: string;
  srcParagraph: string;
  tgtSentence: string;
  tgtParagraph: string;
};

export type AnnotateExample = {
  source: string;
  target: string;
};

export type AnnotateUsageData = {
  explanation: string;
  examples: AnnotateExample[];
};

export type AnnotateResponse = {
  definition: AnnotateDefinitionData;
  usage: AnnotateUsageData;
}

// -------------------------
// Backend return types
// -------------------------
export type AnnotateDomainDataDTO = {
  words: string[];
  spaces: string[];
  sent_ids: number[];
  par_ids: number[];
};

export type AnnotateDefinitionDataDTO = {
  form: string;
  pos_form: string;
  ipa_form: string;

  lemma: string;
  pos_lemma: string;
  ipa_lemma: string;

  gloss: string;

  src_sentence: string;
  src_paragraph: string;
  tgt_sentence: string;
  tgt_paragraph: string;
};

export type AnnotateUsageDataDTO = {
  explanation: string;
  examples: AnnotateExample[];
};

export type AnnotateResponseDTO = {
  definition: AnnotateDefinitionDataDTO;
  usage: AnnotateUsageDataDTO;
};

export type AnnotateRequestDTO = {
  src_lang: string;
  tgt_lang: string;
  ui_lang: string;

  src: AnnotateDomainDataDTO;
  tgt: AnnotateDomainDataDTO;

  tgt_to_src: Record<number, number[]> | Record<string, number[]>;
  tgt_idx: number;
}

// =========================
// ( Annotate Function Helpers )
// -- AnnotateArgs: Defines annotate() API call args
// -- makeAnnotateArgs(): Converts inputs to AnnotateArgs
// -- toAnnotateEntries(): Converts annotate() API call response to DefineEntry / ExplainEntry
// =========================
export type AnnotateArgs = {
  srcLang: string;
  tgtLang: string;
  uiLang: string;

  src: AnnotateDomainData;
  tgt: AnnotateDomainData;

  tgtToSrc: Record<number, number[]>;
  tgtIdx: number;
};

export function makeAnnotateArgs(args: {
  session: ReaderSession;
  srcLang: string;
  tgtLang: string;
  uiLang: string;
  tgtIdx: number;
  isSwapped: boolean;
}): AnnotateArgs {
  const { session, srcLang, tgtLang, uiLang, tgtIdx, isSwapped } = args;

  const srcBlock = isSwapped ? session.alignment.tgt : session.alignment.src;
  const tgtBlock = isSwapped ? session.alignment.src : session.alignment.tgt;
  const targetToSourceMap = isSwapped
    ? session.alignment.align.srcToTgt
    : session.alignment.align.tgtToSrc;

  return {
    srcLang: isSwapped ? tgtLang : srcLang,
    tgtLang: isSwapped ? srcLang : tgtLang,
    uiLang,
    src: {
      words: srcBlock.words,
      spaces: srcBlock.spaces,
      sentIds: srcBlock.sentIds,
      parIds: srcBlock.parIds,
    },
    tgt: {
      words: tgtBlock.words,
      spaces: tgtBlock.spaces,
      sentIds: tgtBlock.sentIds,
      parIds: tgtBlock.parIds,
    },
    tgtToSrc: targetToSourceMap,
    tgtIdx,
  };
}

export function toAnnotateEntries(res: AnnotateResponse): {
  explainData: ExplainEntry;
  defineData: DefineEntry;
} {
  return {
    explainData: {
      explanation: res.usage.explanation,
      examples: res.usage.examples,
    },
    defineData: {
      form: res.definition.form,
      posForm: res.definition.posForm,
      ipaForm: res.definition.ipaForm,

      lemma: res.definition.lemma,
      posLemma: res.definition.posLemma,
      ipaLemma: res.definition.ipaLemma,

      gloss: res.definition.gloss,

      srcSentence: res.definition.srcSentence,
      srcParagraph: res.definition.srcParagraph,
      tgtSentence: res.definition.tgtSentence,
      tgtParagraph: res.definition.tgtParagraph,
    },
  };
}
